updating/update-using-custom-machine-config-pools.adoc

```diff
@@ -0,0 +1,167 @@
+This is required to make sure control plane nodes remain stable.
+----
+
+An OpenShift or Kubernetes cluster is highly available by design and there are many Kuberenetes features (e.g. pod disruption budget, pod affinity, health checks, replicas) which can make the application highly available in case of node failures.
+
+However there might be some secenarios where a controlled rollout of upgrade to the worker nodes is desired. This helps to make sure the mission critical application stays available during the whole update process even if the upgrade process cause a failure. It is also useful in scenarios where we can not have big maintaine window to complete the whole update process. This is can be characterized as a canary release i.e. we can control the rollout of worker nodes by controlling the machine config pools. After the update of the first MCP, the application compatibility can be verfied and then the whole fleet can be updated gradually to the new version.
+
+[NOTE]
+----
+A node can be part of at most one MCP. The machine config operator rolls out update to all the unpaused worker pools simultaneously.
+For example, if maxUnavailable is set to one then 1 node from each MCPs will be updated concurrently.
+
+During the update process of nodes (as per are the  `maxUnavailable` value) in a MCP are cordoned and drained before it is updated.
+Once the update process completes the nodes are un-cordned.
+----
+
+== Workflow of doing a canary release using MCP
+
+. Create MCPs out of the worker pool. The number of nodes in each MCP depends on few factors like maintaince window duration for each MCP, the amount of reserve capacity (extra worker nodes) present.
++
+[NOTE]
+====
+In case of an failure i.e. the MCP with the new version does not work as expected the applications, the extra capacity will help to maintain the quality of service of the applications.
+====
++
+. Pause the MCPs created in the previous step.
++
+. Start the update process. The update process will only update the MCP which contains the master nodes i.e. the control plane.
++
+. Once the control plane update is completed, unpause one MCP. It will update the MCP to the new version of OCP. You can check the progress of the update in the web console (In Administrator view -> Administration -> Cluster settings ) as well as running `$ oc get mcp` CLI command.
++
+[NOTE]
+====
+You can change `maxUnavailable` in a MCP to specify the percentage or the number of machines that can be updating at any given time. default is 1.
+====
++
+. Test if the applications are working as expected on the newly update MCP.
+
+. If you want to gohead with updating the rest of the MCPs then unpause rest of the MCPs one by one till all worker nodes are updated to the new version.
+
+== Steps to create a MCP
+
+. Get the list of worker nodes.
++
+[source,terminal]
+----
+$ oc get nodes | grep -v master
```

Comment:
This `grep` seems brittle.  We should be able to use a role label/annotation to directly select nodes that belong to the `worker` pool.

Reply:
[Label filtering][1] seems to work:

```console
$ oc get -l 'node-role.kubernetes.io/master!=' nodes
NAME                                       STATUS   ROLES    AGE   VERSION
ci-ln-pwnll6b-f76d1-s8t9n-worker-a-s75z4   Ready    worker   20m   v1.21.1+051ac4f
ci-ln-pwnll6b-f76d1-s8t9n-worker-b-dglj2   Ready    worker   20m   v1.21.1+051ac4f
ci-ln-pwnll6b-f76d1-s8t9n-worker-c-lldbm   Ready    worker   20m   v1.21.1+051ac4f
```

[1]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#list-and-watch-filtering

Reply:
Or, if you only need the names:

```console
$ oc get -l 'node-role.kubernetes.io/master!=' -o 'jsonpath={range .items[*]}{.metadata.name}{"\n"}{end}' nodes
ci-ln-pwnll6b-f76d1-s8t9n-worker-a-s75z4
ci-ln-pwnll6b-f76d1-s8t9n-worker-b-dglj2
ci-ln-pwnll6b-f76d1-s8t9n-worker-c-lldbm
```